### ISSUE TITLE
Porting the code to MS MARCO dataset

### ISSUE THREAD

**Issue by burglarhobbit** [HKUST-KnowComp/R-Net#11]
So I have ported this implementation of RNET on the [MS-MARCO](http://www.msmarco.org/) dataset, and the rouge-l scores after training are quite low compared to the reported scores on MS-MARCO in the RNET paper.

[ROUGE-L dev scores](https://github.com/burglarhobbit/R-Net/blob/master/img/dev_rouge_l.png)
[ROUGE-L train scores](https://github.com/burglarhobbit/R-Net/blob/master/img/train_rouge_l.png)

What might have gone wrong?

Here is a link for my implementation reference: [rnet-msmarco](https://github.com/burglarhobbit/R-Net/tree/master/msmarco/rnet)

**Comment by xiongyifan** [comment]
how about the parameters. Did you change them?

**Comment by burglarhobbit** [comment]
I didn't change any parameters apart from the pre-processing which I'm describing below:

1. There is only one query for a passage (several passages, concatenated into one)
2. For each query, there are sometimes one or multiple answers. In that case, I take the first non-empty answer from the list.
3. In SQuAD, there were multiple QA pairs for a passage, which resulted into the `y1s` and `y2s` being appended several times. In MS-MARCO, the size of that list would always remain one. [Code](https://github.com/burglarhobbit/R-Net/blob/master/msmarco/rnet/prepro_msm.py#L195)
4. For answer span generation, I'm using the start and end positions from the longest common subsequence problem as used by the rouge-l metric. [Code](https://github.com/burglarhobbit/R-Net/blob/master/msmarco/rnet/prepro_msm.py#L197)

**Comment by xiongyifan** [comment]
"The answer to the question in MS-MARCO is generated by human and the answer words can not only come from the given text." It comes from the paper of R-Net. I think MS-MARCO dataset is more complicate than SQuAD and the train scores are not good. So maybe the bigger model will be helpful. 
Did you find some project for MS-MARCO? Comparing this project with them. Maybe you can get some ideas.

**Comment by burglarhobbit** [comment]
Yes, the answer words can not only come from the given text. That is why we're using the longest common subsequence (ROUGE-L) to extract a span which consists of those words in the longest order as mentioned in the RNET paper.

In the [R-Net paper, Pg 7](https://www.microsoft.com/en-us/research/wp-content/uploads/2017/05/r-net.pdf), they've reported scores on MS-MARCO with 42.9% accuracy using the ROUGE-L score metric.

> Did you find some project for MS-MARCO

No I did not find any project for MS-MARCO.

> So maybe the bigger model will be helpful.

Can you elaborate on that? Do you mean the original rnet equations when you say "bigger model"?

**Comment by xiongyifan** [comment]
I did not do some works on MS-MARCO. What I mean about the bigger model is something like this.
```
# 75 -> 100
flags.DEFINE_integer("hidden", 75, "Hidden size")
# 0.7 -> 0.9
flags.DEFINE_float("keep_prob", 0.7, "Dropout keep prob in rnn")
```
because the train scores are too lower. Maybe there is the underfitting problem.

**Comment by SeanLee97** [comment]
I am trouble in the marco dataset， Have you solved the problem?  @xiongyifan

**Comment by SeanLee97** [comment]
I am calc the recall values between each passage and question. Then I choose one passage which has the largetest recall value. I use the BiDAF，but it doesn't work. I think I make mistakes in processing the dataset. @xiongyifan

**Comment by xiongyifan** [comment]
@SeanLee97 Sorry, I did not do works on the marco dataset.

**Comment by burglarhobbit** [comment]
what does `config.bucket_range = [40,401,40]` means? As I'm porting it to MS-MARCO and concatenating paragraphs, the para size is becoming around 1500, so do I have to use bucketing and change the `bucket_range` or can I skip it? Please suggest. @xiongyifan @wzhouad

**Comment by xiongyifan** [comment]
if the range is [40,401,40], then it will generate 10 buckets, 0~40, 41 ~ 80, 81 ~ 120 ... 321 ~ 360, 361 ~ 400. 

The code can help you to understand.

```
import tensorflow as tf

buckets = tf.range(40, 401, 40)

t1 = tf.clip_by_value(buckets, 0, 39)
t2 = tf.clip_by_value(buckets, 0, 40)
t3 = tf.clip_by_value(buckets, 0, 41)

sess = tf.Session()

print(sess.run(buckets))
print(sess.run(t1))
print(sess.run(t2))
print(sess.run(t3))
```

**Comment by burglarhobbit** [comment]
Thank you very much. I would also like to know whether bucketing would improve the scores or just the training efficiency?

Also, does heavy padding [for example, if one query has only one para as its context, but another query has 10 para as its context, so I would have to pad the data of the first query to the length of extra 9*para-lengths] affect the scores much? I'm strictly talking in terms of the MS-MARCO dataset which is why I'm explaining and elaborating the situations.

Another thing I have observed in the code [model.py Line#25](https://github.com/HKUST-KnowComp/R-Net/blob/master/model.py#L25) is that you trim the [context and question] indexes which are having zeros from the back using the cast->tf.bool->sum->maxlen procedure. So the embedding/word vector gradients that pass through the network have the dimensions equivalent to the actual para/ques length, and not the padded length, right? Please clarify.

**Comment by xiongyifan** [comment]
```
# In the README.md, Detailed Implementaion, the last sentence.
The bucketing method can speedup the training, but will lower the F1 score by 0.3%.
```

I think in the beginning the speed is important, So the bucketing method is a good choice. Finally close it for more score. 

The second. I am not so clear about your problem. I try to describe it.
the code from 24~34 is only for the performance. after it the batch will have max length in the batch rather than in the config.py. 

Here is a another important thing about the gradients. 
```
 _, (state_fw, state_bw) = tf.nn.bidirectional_dynamic_rnn( cell_fw, cell_bw, ch_emb, self.ch_len, dtype=tf.float32)
```
When you use bidirectional_dynamic_rnn, There is a parameter sequence_length. 
```
# the comment about it from dynamic_rnn function.
sequence_length: (optional) An int32/int64 vector sized `[batch_size]`.
  Used to copy-through state and zero-out outputs when past a batch
  element's sequence length.  So it's more for correctness than performance.
```
`correctness ` is a key word. The meaning is the final state is the last nonzero position depending on the sequence_length parameter. So the padding will not affect the gradients.
